dali/operators/audio/nonsilence_op_gpu.cu

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/reduce/find_region.cuh"
+#include "dali/kernels/reduce/reduce_gpu.h"
+#include "dali/kernels/signal/decibel/decibel_calculator.h"
+#include "dali/kernels/signal/moving_mean_square_gpu.h"
+#include "dali/operators/audio/nonsilence_op.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+
+namespace {
+
+/**
+ * @brief Threshold predicate for the nonsilent region.
+ */
+template <typename T>
+struct NotLess {
+  T value_;
+
+  DALI_HOST_DEV DALI_FORCEINLINE bool operator()(T x) const noexcept {
+    return x >= value_;
+  }
+};
+
+struct InitPredicateSampleArgs {
+  NotLess<float> *predicate;
+  float *ref_pow;
+  float cutoff_db;
+};
+
+__global__ void InitPredicates(InitPredicateSampleArgs *sample_params, int nsamples) {
+  for (int idx = threadIdx.x + blockIdx.x * blockDim.x; idx < nsamples;
+       idx += blockDim.x * gridDim.x) {
+    auto &params = sample_params[idx];
+    kernels::signal::DecibelToMagnitude<float> db2mag(10.f, *params.ref_pow);
+    *(params.predicate) = {db2mag(params.cutoff_db)};
+  }
+}
+
+template <typename Idx>
+struct NonsilentRegionPostprocessArgs {
+  Idx *begin;
+  Idx *len;
+  const i64vec2 *region;
+  Idx input_len;
+};
+
+template <typename Idx = int64_t>
+__global__ void NonsilentRegionPostprocess(NonsilentRegionPostprocessArgs<Idx> *args, int nsamples,
+                                           int window) {
+  for (int idx = threadIdx.x + blockIdx.x * blockDim.x; idx < nsamples;
+       idx += blockDim.x * gridDim.x) {
+    auto &sample_args = args[idx];
+    auto region = *(sample_args.region);
+    // Begin it's calculated as first - window, meaning that we start counting from the
+    // beginning of the sliding window. The lenght of the region needs to be increased by the same
+    // factor. We also limit the region to the bounds of the input.
+    Idx start = cuda_max(static_cast<Idx>(region.x - window), Idx(0));
+    *(sample_args.begin) = start;
+    *(sample_args.len) = cuda_min(static_cast<Idx>(region.y + window),
```

Comment:
Is the result of finding a region a `start, length` pair? I think it should be `start, end` and we could do the additional postprocessing here. The `start, length` representation may be nice when typing in by hand, but internally, it's almost always better to work with `start, end`.